File: pcai.py
import requests
import json
import web3

def scan_activity(activity_id):
    # Get the activity data
    response = requests.get(f"https://api.metamask.io/activities/{activity_id}")
    data = json.loads(response.content)

    # Get the receiving wallet addresses
    receiving_addresses = []
    for transaction in data["transactions"]:
        receiving_addresses.append(transaction["to"])

    # Get the sending wallet address
    sending_address = data["from"]

    # Check if the receiving wallet has the sending wallet address
    if sending_address in receiving_addresses:
        # Allow the transaction
        return True
    else:
        # Block the transaction and prompt the user to whitelist the sending wallet
        print("The transaction has been blocked. Please enter the sending wallet address to whitelist it: ")
        sending_wallet_address = input()
        web3.eth.defaultAccount = sending_wallet_address
        return False

def main():
    # Get the activity ID
    activity_id = input("Enter the activity ID: ")

    # Scan the activity
    allowed = scan_activity(activity_id)

    # If the transaction is allowed, do nothing
    if allowed:
        pass

    # If the transaction is blocked, print a message and exit
    else:
        print("The transaction has been blocked.")
        exit()

if __name__ == "__main__":
    main()

File: pcai.js
function scanActivity(activityId) {
  const response = await fetch(`https://api.metamask.io/activities/${activityId}`);
  const data = await response.json();

  const receivingAddresses = [];
  for (const transaction of data.transactions) {
    receivingAddresses.push(transaction.to);
  }

  const sendingAddress = data.from;

  return sendingAddress in receivingAddresses ? true : false;
}

function signTransaction(data, provider) {
  const provider = detectEthereumProvider();

  const signature = provider.personalSign(data);

  return signature;
}

function signFile(filePath, provider) {
  const data = await fs.readFileSync(filePath, "rb");

  const signature = signTransaction(data, provider);

  await fs.writeFileSync(filePath + ".sig", signature);
}

function uploadSignature(signature, decentralizedStorageNetwork) {
  const ipfsClient = decentralizedStorageNetwork.getIpfsClient();

  const ipfsHash = await ipfsClient.add(signature);

  return ipfsHash;
}

async function main() {
  const activityId = await prompt("Enter the activity ID: ");

  const allowed = await scanActivity(activityId);

  if (allowed) {
    const provider = web3.eth.get_default_provider();

    provider.allowSameNetworkTransactions = true;

    const networkId = await provider.getNetworkId();

    const signature = await signTransaction(data, provider);

    await fs.writeFileSync("signature.sig", signature);

    const ipfsHash = await uploadSignature(signature, IPFS);

    await fs.appendFileSync("whitelist.txt", sendingWalletAddress + "\n");
  } else {
    const sendingWalletAddress = await prompt("Enter the sending wallet address to whitelist: ");

    await fs.appendFileSync("whitelist.txt", sendingWalletAddress + "\n");
  }
}

if (__name__ == "__main__") {
  main();
}
